src/softmax.jl

```diff
@@ -26,9 +28,9 @@ julia> softmax([1, 2, 3])
 See also [`logsoftmax`](@ref).
 """
 function softmax(xs::AbstractArray; dims=1)
-    max_ = maximum(xs, dims=dims)
-    exp_ = exp.(xs .- max_)
-    exp_ ./ sum(exp_, dims=dims)
+    max_ = vreduce(max, xs, dims=dims)
+    exp_ = SLEEF.exp.(xs .- max_)
+    exp_ ./ vreduce(+, exp_, dims=dims)
```

Comment:
also here, these operations won't support generic `AbstractArray`s. This is very annoying, but if we want this we'll have to create a copy of the original softmax specialized on Array`

Reply:
Dispatch inside LoopVectorization functions and macros should automatically be hitting `SLEEFPirates.exp`. Is that not happening?

Reply:
`SLEEF.exp.(...)` can handle generic `AbstractArrays` and generate vectorized code just as `@avx`. I will change that to be `@avx exp.(...)` or `vmap(exp, ...)` later, but `vmap` `vreduce` and `@avx` may not work on generic  `AbstractArrays`. @chriselrod Do we need to dispatch non `Array` type to Base methods in `LoopVectorization`?

Reply:
Another way is to add seperate methods in NNlib instead of LoopVectorization to deal with different  `AbstractArrays`. What do you think? @CarloLucibello @chriselrod

Reply:
LoopVectorization should be "compatible" with generic `AbstractArrays` in that it calls `check_args` on them:
```julia
julia> @macroexpand @avx for i ∈ eachindex(x, y)
           s += x[i] * y[i]
       end
quote
    begin
        var"##loopi#366" = LoopVectorization.maybestaticrange(eachindex(x, y))
        var"##i_loop_lower_bound#367" = LoopVectorization.maybestaticfirst(var"##loopi#366")
        var"##i_loop_upper_bound#368" = LoopVectorization.maybestaticlast(var"##loopi#366")
    end
    if LoopVectorization.check_args(x, y)
        var"##vptr##_x" = LoopVectorization.stridedpointer(x)
        var"##vptr##_y" = LoopVectorization.stridedpointer(y)
        local var"##s_0"
        begin
            $(Expr(:gc_preserve, :(var"##s_0" = begin
          begin
              var"##Tloopeltype##" = promote_type(eltype(x), eltype(y))
              var"##Wvecwidth##" = LoopVectorization.pick_vector_width_val(eltype(x), eltype(y))
          end
          LoopVectorization._avx_!(Val{(0, 0, 0, LoopVectorization.unwrap(var"##Wvecwidth##"))}(), Tuple{:LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000001, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x01, 0x01), :LoopVectorization, :getindex, LoopVectorization.OperationStruct(0x0000000000000001, 0x0000000000000000, 0x0000000000000000, 0x0000000000000000, LoopVectorization.memload, 0x02, 0x02), :LoopVectorization, :LOOPCONSTANTINSTRUCTION, LoopVectorization.OperationStruct(0x0000000000000000, 0x0000000000000000, 0x0000000000000001, 0x0000000000000000, LoopVectorization.constant, 0x00, 0x03), :LoopVectorization, :vfmadd_fast, LoopVectorization.OperationStruct(0x0000000000000001, 0x0000000000000001, 0x0000000000000000, 0x0000000000010203, LoopVectorization.compute, 0x00, 0x03)}, Tuple{LoopVectorization.ArrayRefStruct{:x,Symbol("##vptr##_x")}(0x0000000000000001, 0x0000000000000001, 0x0000000000000000), LoopVectorization.ArrayRefStruct{:y,Symbol("##vptr##_y")}(0x0000000000000001, 0x0000000000000001, 0x0000000000000000)}, Tuple{0, Tuple{4}, Tuple{3}, Tuple{}, Tuple{}, Tuple{}, Tuple{}}, Tuple{:i}, (var"##i_loop_lower_bound#367":var"##i_loop_upper_bound#368",), var"##vptr##_x", var"##vptr##_y", s)
      end), :x, :y))
        end
        s = LoopVectorization.reduced_add(var"##s_0", s)
    else
        $(Expr(:inbounds, true))
        local var"#20#val" = for i = eachindex(x, y)
                    #= REPL[3]:2 =#
                    s = Base.FastMath.add_fast(s, Base.FastMath.mul_fast(x[i], y[i]))
                end
        $(Expr(:inbounds, :pop))
        var"#20#val"
    end
end
```
If `check_args` returns true, it will call `_avx_!`, which will use LoopVectorization to optimize the loop.
If `check_args` returns false, it'll use `@inbounds @fastmath`.
```julia
julia> using LinearAlgebra

julia> LoopVectorization.check_args(rand(3,3))
true

julia> LoopVectorization.check_args(Symmetric(rand(3,3)))
false
```

Reply:
This is for loops. For broadcasts, it dispatches on the return type. So it's probably better to have explicit checks, until I switch to handling it more similarly to loops.

Reply:
`SLEEF.exp` in a broadcast is probably fine. For `Float32`, I found it was faster for long loops than using LoopVectorization, because when LLVM vectorized it, it unrolled it and interleaved all the individual instructions of the `exp`, letting out of order execution work to compute a few in parallel (on top of the SIMD).

Reply:
Well, that's what I remembered. Not what I'm seeing at the moment:
```julia
julia> using SLEEFPirates

julia> x128 = rand(Float32, 128); y128 = similar(x128);

julia> x127 = rand(Float32, 127); y127 = similar(x127);

julia> @btime @. $y128 = SLEEFPirates.exp($x128);
  42.911 ns (0 allocations: 0 bytes)

julia> @btime @avx @. $y128 = SLEEFPirates.exp($x128);
  34.418 ns (0 allocations: 0 bytes)

julia> @btime @avx @. $y128 = exp($x128);
  34.660 ns (0 allocations: 0 bytes)

julia> @btime @. $y127 = SLEEFPirates.exp($x127);
  106.803 ns (0 allocations: 0 bytes)

julia> @btime @avx @. $y127 = exp($x127);
  34.787 ns (0 allocations: 0 bytes)

julia> x512 = rand(Float32, 512); y512 = similar(x512);

julia> @btime @. $y512 = SLEEFPirates.exp($x512);
  139.423 ns (0 allocations: 0 bytes)

julia> @btime @avx @. $y512 = exp($x512);
  125.525 ns (0 allocations: 0 bytes)

julia> x1024 = rand(Float32, 1024); y1024 = similar(x1024);

julia> @btime @. $y1024 = SLEEFPirates.exp($x1024);
  276.671 ns (0 allocations: 0 bytes)

julia> @btime @avx @. $y1024 = exp($x1024);
  255.173 ns (0 allocations: 0 bytes)
```

Reply:
In the latest commit, I found that `vmap` and `vreduce` don't work for `StaticArrays` while the `@avx` loops work. So I created functions like `fsum` `fmaximum` and `fmap` to get around this. If you'd like to modify those functions in `LoopVectorization`, I will remove those ugly definitions.

Reply:
The negligible performance degragation of migrating from `SLEEFPirates.exp.(...)` to `vmap(exp, ...)` seems fine to me.